contracts/tvl-staking/RibbonDualStakingRewards.sol

```diff
@@ -0,0 +1,355 @@
+    uint256 totalSupply_ = _totalSupply;
+    if (totalSupply_ == 0) {
+      return (_rewardPerTokenStored.token0, _rewardPerTokenStored.token1);
+    }
+
+    Rewards memory _rewardRate = rewardRate;
+    uint256 _lastTimeRewardApplicable = lastTimeRewardApplicable().sub(
+      lastUpdateTime
+    );
+    return (
+      uint256(_rewardPerTokenStored.token0).add(
+        _lastTimeRewardApplicable.mul(_rewardRate.token0).mul(1e18).div(
+          totalSupply_
+        )
+      ),
+      uint256(_rewardPerTokenStored.token1).add(
+        _lastTimeRewardApplicable.mul(_rewardRate.token1).mul(1e18).div(
+          totalSupply_
+        )
+      )
+    );
+  }
+
+  function earned(address account)
+    public
+    view
+    override
+    returns (uint256, uint256)
+  {
+    (uint256 rewardPerToken0, uint256 rewardPerToken1) = rewardPerToken();
+    uint256 balance = _balances[account];
+    Rewards memory _userRewardPerTokenPaid = userRewardPerTokenPaid[account];
+    Rewards memory _rewards = rewards[account];
+    return (
+      balance
+        .mul(rewardPerToken0.sub(_userRewardPerTokenPaid.token0))
+        .div(1e18)
+        .add(_rewards.token0),
+      balance
+        .mul(rewardPerToken1.sub(_userRewardPerTokenPaid.token1))
+        .div(1e18)
+        .add(_rewards.token1)
+    );
+  }
+
+  function getRewardForDuration()
+    external
+    view
+    override
+    returns (uint256, uint256)
+  {
+    Rewards memory _rewardRate = rewardRate;
+    uint256 _rewardsDuration = rewardsDuration;
+    return (
+      uint256(_rewardRate.token0).mul(_rewardsDuration),
+      uint256(_rewardRate.token1).mul(_rewardsDuration)
+    );
+  }
+
+  /* ========== MUTATIVE FUNCTIONS ========== */
+
+  function stake(uint256 amount) external override {
+    stakeFor(amount, msg.sender);
+  }
+
+  function stakeFor(uint256 amount, address account)
+    public
+    override
+    nonReentrant
+    notPaused
+    updateReward(account)
+  {
+    require(amount > 0, "Cannot stake 0");
+    _totalSupply = _totalSupply.add(amount);
+    _balances[account] = _balances[account].add(amount);
+    stakingToken.safeTransferFrom(msg.sender, address(this), amount);
+    emit Staked(account, msg.sender, amount);
+  }
+
+  function withdraw(uint256 amount)
+    public
+    override
+    nonReentrant
+    updateReward(msg.sender)
+  {
+    require(amount > 0, "Cannot withdraw 0");
+    _totalSupply = _totalSupply.sub(amount);
+    _balances[msg.sender] = _balances[msg.sender].sub(amount);
+    if (block.timestamp < periodFinish.add(1 days)) {
+      delete rewards[msg.sender];
+    }
+    stakingToken.safeTransfer(msg.sender, amount);
+    emit Withdrawn(msg.sender, amount);
+  }
+
+  function getReward() public override nonReentrant updateReward(msg.sender) {
+    if (block.timestamp >= periodFinish.add(1 days)) {
+      Rewards memory _rewards = rewards[msg.sender];
+      if (_rewards.token0 > 0) {
+        rewardsToken0.safeTransfer(
```

Comment:
We could cache the `rewardsToken0` storage read in a variable here since we access it a couple of times

Reply:
since the staking/rewards token vars are now immutable it doesn't result in a storage read, hence no need to cache

Reply:
oh sorry, forgot about that